src/main/java/org/terasology/metalrenegades/world/dynamic/FloraProvider.java

```diff
@@ -0,0 +1,124 @@
+// Copyright 2021 The Terasology Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+package org.terasology.metalrenegades.world.dynamic;
+
+import com.google.common.base.Predicate;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Lists;
+import org.joml.Vector3i;
+import org.terasology.biomesAPI.Biome;
+import org.terasology.core.world.CoreBiome;
+import org.terasology.core.world.generator.facetProviders.PositionFilters;
+import org.terasology.core.world.generator.facetProviders.SurfaceObjectProvider;
+import org.terasology.core.world.generator.facets.BiomeFacet;
+import org.terasology.core.world.generator.facets.FloraFacet;
+import org.terasology.core.world.generator.rasterizers.FloraType;
+import org.terasology.engine.entitySystem.Component;
+import org.terasology.engine.utilities.procedural.Noise;
+import org.terasology.engine.utilities.procedural.WhiteNoise;
+import org.terasology.engine.world.generation.ConfigurableFacetProvider;
+import org.terasology.engine.world.generation.Facet;
+import org.terasology.engine.world.generation.FacetBorder;
+import org.terasology.engine.world.generation.GeneratingRegion;
+import org.terasology.engine.world.generation.Produces;
+import org.terasology.engine.world.generation.Requires;
+import org.terasology.engine.world.generation.facets.SeaLevelFacet;
+import org.terasology.engine.world.generation.facets.SurfacesFacet;
+import org.terasology.nui.properties.Range;
+
+import java.util.List;
+import java.util.Map;
+
+/**
+ * For the most part, this is just a copy of DefaultFloraProvider from CoreWorlds, extended with the new biomes.
+ */
+@Produces(FloraFacet.class)
+@Requires({
+        @Facet(SeaLevelFacet.class),
+        @Facet(value = SurfacesFacet.class, border = @FacetBorder(bottom = 1)),
+        @Facet(BiomeFacet.class)
+})
+public class FloraProvider extends SurfaceObjectProvider<Biome, FloraType> implements ConfigurableFacetProvider {
+    private Noise densityNoiseGen;
+
+    private Configuration configuration = new Configuration();
+
+    private Map<FloraType, Float> typeProbs = ImmutableMap.of(
+            FloraType.GRASS, 0.85f,
+            FloraType.FLOWER, 0.1f,
+            FloraType.MUSHROOM, 0.05f);
+
+    private Map<Biome, Float> biomeProbs = ImmutableMap.<Biome, Float>builder()
+            .put(MRBiome.RIVER, 0.2f)
+            .put(MRBiome.STEPPE, 0.1f)
+            .put(MRBiome.ROCKY, 0.001f)
+            .put(MRBiome.SCRUBLAND, 0.05f)
+            .put(CoreBiome.DESERT, 0.001f).build();
```

Comment:
Skimming through the implementation of FloraProvider, I'm guessing the first set of numbers tell how likely each type of vegetation is and the second set of numbers tell how likely vegetation will be able to grow on that biome. The probability distribution for every biome is the same. Does this mean that the proportion of flowers in a desert would be the same as the proportion of flowers in a river biome? I think it might be better to have a different set of probabilities for different biomes. For example, cactii would be more common in deserts than in rivers.(even though the probability of any vegetation growing would be much lesser in a desert)

Reply:
Yes, the probabilities for each type are the same for each biome. However, desert is somewhat different; cacti are actually treated as trees and handled by the DefaultTreeProvider, which only spawns them in deserts, and mushrooms are hardcoded not to generate in deserts. So the probabilities for the desert biome are actually:

- Cactus: 0.04
- Grass: 0.0085
- Flower: 0.0001

which seems pretty reasonable to me. That said, a way to change type probabilities per biome might be useful in the future.